File: ios/FieldAgentModule.swift
import CoreLocation
import ExpoModulesCore
import UIKit
import UserNotifications

/// iOS surface of expo-field-agent.
///
/// The whole API exists here. Capabilities the platform does not have return an
/// explicit value — `false`, `"unsupported"` — and never throw: a host must be
/// able to write one code path for both platforms.
public class FieldAgentModule: Module {

  private let tracker = FieldAgentTracker.shared
  private let alerts = FieldAgentAlerts.shared

  public func definition() -> ModuleDefinition {
    Name("FieldAgent")

    Events("position", "sent", "error", "alert", "bubblePress")

    OnCreate {
      self.tracker.onPosition = { [weak self] payload in self?.sendEvent("position", payload) }
      self.tracker.onSent = { [weak self] count, queued in
        self?.sendEvent("sent", ["count": count, "queued": queued])
      }
      self.tracker.onError = { [weak self] code, message in
        self?.sendEvent("error", ["code": code, "message": message])
      }
      self.alerts.onAlert = { [weak self] alert in self?.sendEvent("alert", alert) }
      // A significant location change can relaunch the app straight into this
      // module; resuming here is what makes that relaunch worth anything.
      self.tracker.resumeIfWanted()
    }

    OnDestroy {
      self.tracker.onPosition = nil
      self.tracker.onSent = nil
      self.tracker.onError = nil
      self.alerts.onAlert = nil
    }

    // MARK: Permissions

    AsyncFunction("getPermissions") { (promise: Promise) in
      self.readPermissions { promise.resolve($0) }
    }

    AsyncFunction("requestPermissions") { (skip: [String], promise: Promise) in
      let skipped = Set(skip)

      let finish = { self.readPermissions { promise.resolve($0) } }

      let askNotifications = {
        guard !skipped.contains("notifications") else { return finish() }
        var options: UNAuthorizationOptions = [.alert, .sound, .badge]
        if FieldAgentConfig.current.criticalAlerts { options.insert(.criticalAlert) }
        UNUserNotificationCenter.current().requestAuthorization(options: options) { _, _ in
          DispatchQueue.main.async { finish() }
        }
      }

      guard !(skipped.contains("location") && skipped.contains("backgroundLocation")) else {
        return askNotifications()
      }

      // iOS imposes the same two-step ladder as Android: when-in-use first,
      // "always" only afterwards, and only the system may show the second prompt.
      self.requestLocation(wantsAlways: !skipped.contains("backgroundLocation")) {
        askNotifications()
      }
    }

    AsyncFunction("openSettings") { (_: String) -> Void in
      // iOS has exactly one destination for all of them.
      guard let url = URL(string: UIApplication.openSettingsURLString) else { return }
      DispatchQueue.main.async { UIApplication.shared.open(url) }
    }

    // MARK: Tracking

    AsyncFunction("start") { (options: [String: Any]?) -> Void in
      FieldAgentConfig.setOverrides(options)
      try self.tracker.start()
    }

    AsyncFunction("stop") { () -> Void in self.tracker.stop() }

    AsyncFunction("isRunning") { () -> Bool in self.tracker.isRunning }

    AsyncFunction("setAuthHeader") { (value: String?) -> Void in self.tracker.setAuthHeader(value) }

    AsyncFunction("setIntervalSeconds") { (seconds: Double) -> Void in self.tracker.setInterval(seconds) }

    AsyncFunction("flush") { (promise: Promise) in
      self.tracker.flush { sent, queued in
        promise.resolve(["sent": sent, "queued": queued])
      }
    }

    AsyncFunction("getState") { () -> [String: Any] in self.tracker.state() }

    // MARK: Bubble — no such thing on iOS, and there never will be

    AsyncFunction("showBubble") { () -> Bool in false }
    AsyncFunction("hideBubble") { () -> Void in }
    AsyncFunction("setBubbleState") { (_: String, _: String?) -> Void in }

    // MARK: Alert

    AsyncFunction("triggerAlert") { (payload: [String: Any]) -> Void in
      guard let title = payload["title"] as? String, !title.isEmpty else {
        throw FieldAgentError.invalidPayload
      }
      let data = payload["data"] as? [String: Any]
      let silent = (data?["silent"] as? Bool) == true
      DispatchQueue.main.async {
        self.alerts.trigger(
          title: title,
          body: payload["body"] as? String,
          data: data,
          silent: silent,
          tag: payload["tag"] as? String,
          channelId: payload["channelId"] as? String
        )
      }
    }

    AsyncFunction("dismissAlert") { () -> Void in
      DispatchQueue.main.async { self.alerts.dismiss() }
    }

    AsyncFunction("setAlertSound") { (enabled: Bool) -> Void in self.alerts.soundEnabled = enabled }

    Function("getPendingAlertSync") { () -> [String: Any]? in self.alerts.pending() }
  }

  // MARK: - Helpers

  private func readPermissions(_ completion: @escaping ([String: String]) -> Void) {
    let location: String
    let background: String

    switch tracker.authorizationStatus {
    case .authorizedAlways:
      location = "granted"
      background = "granted"
    case .authorizedWhenInUse:
      location = "granted"
      background = "denied"
    case .denied, .restricted:
      location = "denied"
      background = "denied"
    default:
      location = "undetermined"
      background = "undetermined"
    }

    UNUserNotificationCenter.current().getNotificationSettings { settings in
      let notifications: String
      switch settings.authorizationStatus {
      case .authorized, .provisional, .ephemeral: notifications = "granted"
      case .denied: notifications = "denied"
      default: notifications = "undetermined"
      }

      completion([
        "location": location,
        "backgroundLocation": background,
        "notifications": notifications,
        // No overlay window, no battery allow-list, no notification policy
        // access, no full-screen intent, no manufacturer autostart on iOS.
        "overlay": "unsupported",
        "batteryUnrestricted": "unsupported",
        "dndAccess": "unsupported",
        "fullScreenIntent": "unsupported",
        "autostart": "unsupported"
      ])
    }
  }

  private func requestLocation(wantsAlways: Bool, completion: @escaping () -> Void) {
    // The system prompt can be dismissed without ever changing the status, so
    // every wait carries a deadline: a permission call that never resolves is a
    // host stuck on a spinner.
    var settled = false
    let once = {
      guard !settled else { return }
      settled = true
      completion()
    }
    DispatchQueue.main.asyncAfter(deadline: .now() + 60) { once() }

    switch tracker.authorizationStatus {
    case .notDetermined:
      tracker.onAuthorizationChange = { [weak self] in
        guard let self, wantsAlways, self.tracker.authorizationStatus == .authorizedWhenInUse else {
          return once()
        }
        // Second step: iOS offers "Always" only after when-in-use was granted.
        self.tracker.onAuthorizationChange = { once() }
        self.tracker.requestAlways()
      }
      tracker.requestWhenInUse()

    case .authorizedWhenInUse where wantsAlways:
      tracker.onAuthorizationChange = { once() }
      tracker.requestAlways()

    default:
      once()
    }
  }
}

File: android/src/main/java/expo/modules/fieldagent/FieldAgentModule.kt
package expo.modules.fieldagent

import android.Manifest
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.provider.Settings
import androidx.core.content.ContextCompat
import expo.modules.interfaces.permissions.PermissionsResponseListener
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlin.coroutines.resume
import kotlinx.coroutines.suspendCancellableCoroutine
import org.json.JSONObject

private const val RC_BACKGROUND_LOCATION = 0xFA20
private const val RC_OVERLAY = 0xFA21
private const val RC_DND = 0xFA22
private const val RC_BATTERY = 0xFA23
private const val RC_FULL_SCREEN = 0xFA24
private const val RC_AUTOSTART = 0xFA25

class FieldAgentModule : Module() {

    private val context: Context
        get() = appContext.reactContext?.applicationContext ?: throw Exceptions.ReactContextLost()

    private val main = Handler(Looper.getMainLooper())

    private var settingsRequestCode = 0
    private var settingsContinuation: ((Unit) -> Unit)? = null

    override fun definition() = ModuleDefinition {
        Name("FieldAgent")

        Events("position", "sent", "error", "alert", "bubblePress")

        OnCreate {
            Bus.listener = { name, payload -> runCatching { this@FieldAgentModule.sendEvent(name, payload) } }
            // A process killed mid-alert would otherwise leave the user's alarm
            // volume pinned at maximum with no idea why.
            Alerts.recover(context)
        }

        OnDestroy {
            Bus.listener = null
        }

        OnActivityResult { _, payload ->
            if (payload.requestCode == settingsRequestCode) {
                val resume = settingsContinuation
                settingsContinuation = null
                resume?.invoke(Unit)
            }
        }

        // --- Permissions ---------------------------------------------------

        AsyncFunction("getPermissions") { permissionsBundle() }

        AsyncFunction("requestPermissions") Coroutine { skip: List<String> ->
            requestLadder(skip.toSet())
            permissionsBundle()
        }

        AsyncFunction("openSettings") Coroutine { which: String ->
            openSettingsFor(which)
        }

        // --- Tracking ------------------------------------------------------

        AsyncFunction("start") { options: Map<String, Any?>? ->
            applyOverrides(options)
            val config = Config.get(context)
            if (config.tracking.url.isNullOrEmpty()) {
                throw IllegalStateException(
                    "tracking.url est absente : renseigne-la dans app.json ou passe start({ url })."
                )
            }
            if (ContextCompat.checkSelfPermission(context, Manifest.permission.ACCESS_FINE_LOCATION) !=
                android.content.pm.PackageManager.PERMISSION_GRANTED
            ) {
                throw IllegalStateException("ACCESS_FINE_LOCATION n'est pas accordee : appelle requestPermissions().")
            }

            // Idempotent: the persisted intention is set first so a boot or a
            // watchdog tick resumes even if this very start is refused.
            Prefs.setDesiredRunning(context, true)
            TrackingService.request(context, "js")
            Watchdog.arm(context)
        }

        AsyncFunction("stop") {
            TrackingService.stop(context)
        }

        AsyncFunction("isRunning") { TrackingService.isRunning }

        AsyncFunction("setAuthHeader") { value: String? ->
            if (value == null) {
                Prefs.putString(context, Prefs.AUTH_HEADER, null)
            } else {
                val encrypted = Crypto.encrypt(value)
                    ?: throw IllegalStateException("Le keystore a refuse de chiffrer l'en-tete d'authentification.")
                Prefs.putString(context, Prefs.AUTH_HEADER, encrypted)
            }
        }

        AsyncFunction("setIntervalSeconds") { seconds: Int ->
            require(seconds >= 1) { "setInterval attend un nombre de secondes >= 1." }
            TrackingService.send(
                context,
                TrackingService.ACTION_SET_INTERVAL,
                Bundle().apply { putInt(TrackingService.EXTRA_INTERVAL, seconds) }
            )
        }

        // Runs off the JS thread on the module queue, which is what makes a
        // blocking flush acceptable here.
        AsyncFunction("flush") {
            val (sent, queued) = Outbox.flush(context)
            Bundle().apply {
                putInt("sent", sent)
                putInt("queued", queued)
            }
        }

        AsyncFunction("getState") {
            val preferences = Prefs.of(context)
            val lastFix = preferences.getLong(Prefs.LAST_FIX_AT, 0L)
            val lastSent = preferences.getLong(Prefs.LAST_SENT_AT, 0L)
            // A Map rather than a Bundle: the contract says these are nullable,
            // and a Bundle can only omit a key, which reads as undefined in JS.
            mapOf(
                "running" to TrackingService.isRunning,
                "queued" to Outbox.size(context),
                "lastFixAt" to if (lastFix > 0) lastFix.toDouble() else null,
                "lastSentAt" to if (lastSent > 0) lastSent.toDouble() else null,
                "lastError" to preferences.getString(Prefs.LAST_ERROR, null)
            )
        }

        // --- Bubble --------------------------------------------------------

        AsyncFunction("showBubble") { Bubble.show(context) }

        AsyncFunction("hideBubble") { Bubble.hide(context) }

        AsyncFunction("setBubbleState") { state: String, text: String? ->
            Bubble.setState(context, state, text)
        }

        // --- Alert ---------------------------------------------------------

        AsyncFunction("triggerAlert") { payload: Map<String, Any?> ->
            val title = payload["title"] as? String
                ?: throw IllegalArgumentException("triggerAlert attend { title: string }.")
            val body = payload["body"] as? String
            @Suppress("UNCHECKED_CAST")
            val data = (payload["data"] as? Map<String, Any?>)?.let { JSONObject(it).toString() }
            val silent = ((payload["data"] as? Map<*, *>)?.get("silent") as? Boolean) == true

            val tag = payload["tag"] as? String
            val channelId = payload["channelId"] as? String

            // MediaPlayer, torch and the TTL handler all belong to the main thread.
            main.post { Alerts.trigger(context, title, body, data, silent, tag, channelId) }
        }

        AsyncFunction("dismissAlert") {
            main.post { Alerts.dismiss(context) }
        }

        AsyncFunction("setAlertSound") { enabled: Boolean ->
            Prefs.of(context).edit().putBoolean(Prefs.ALERT_SOUND_ENABLED, enabled).apply()
        }

        Function("getPendingAlertSync") {
            Alerts.peek(context)?.toBundle()
        }
    }

    // --- Permission plumbing ----------------------------------------------

    private fun runtimeState(permission: String): String {
        if (ContextCompat.checkSelfPermission(context, permission) ==
            android.content.pm.PackageManager.PERMISSION_GRANTED
        ) {
            return "granted"
        }
        // "Refused" and "never asked" are different things: only the first must
        // stop the host from asking again.
        return if (Prefs.of(context).getBoolean("asked_$permission", false)) "denied" else "undetermined"
    }

    private fun markAsked(vararg permissions: String) {
        val editor = Prefs.of(context).edit()
        permissions.forEach { editor.putBoolean("asked_$it", true) }
        editor.apply()
    }

    private fun permissionsBundle(): Bundle = Bundle().apply {
        putString("location", runtimeState(Manifest.permission.ACCESS_FINE_LOCATION))
        putString(
            "backgroundLocation",
            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) "granted"
            else runtimeState(Manifest.permission.ACCESS_BACKGROUND_LOCATION)
        )
        putString(
            "notifications",
            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.TIRAMISU) {
                if (Alerts.hasNotificationPermission(context)) "granted" else "denied"
            } else {
                runtimeState(Manifest.permission.POST_NOTIFICATIONS)
            }
        )
        putString("overlay", if (Overlay.isGranted(context)) "granted" else "denied")
        putString("batteryUnrestricted", if (Power.isUnrestricted(context)) "granted" else "denied")
        putString("dndAccess", if (Alerts.hasDndAccess(context)) "granted" else "denied")
        // Android 14 gates the full-screen intent behind a special access; without
        // it the locked-screen path silently degrades to a heads-up notification.
        putString(
            "fullScreenIntent",
            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.UPSIDE_DOWN_CAKE) "granted"
            else if (Alerts.canUseFullScreenIntent(context)) "granted" else "denied"
        )
        // Only meaningful when the host opted into the bridge; otherwise the
        // service is not even in the manifest and the access grants nothing.
        putString(
            "notificationAccess",
            when {
                !NotificationBridge.isEnabled(context) -> "unsupported"
                NotificationBridge.isGranted(context) -> "granted"
                else -> "denied"
            }
        )
        putString(
            "autostart",
            when {
                !Power.hasManufacturerScreen(context) -> "unsupported"
                Power.isConfirmed(context) -> "granted"
                else -> "undetermined"
            }
        )
    }

    private suspend fun requestLadder(skip: Set<String>) {
        // The order is Android's, not ours: foreground location before
        // background, notifications before anything that relies on them.
        if ("location" !in skip) {
            ask(Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.ACCESS_COARSE_LOCATION)
        }

        if ("notifications" !in skip && Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            ask(Manifest.permission.POST_NOTIFICATIONS)
        }

        if ("backgroundLocation" !in skip &&
            Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q &&
            runtimeState(Manifest.permission.ACCESS_FINE_LOCATION) == "granted" &&
            runtimeState(Manifest.permission.ACCESS_BACKGROUND_LOCATION) != "granted"
        ) {
            // Two steps, imposed by the platform. On API 29 the dialog can still
            // grant it; from API 30 only the settings page can, so a single
            // request never returns granted and the host must explain why in
            // between — Google Play requires that disclosure anyway.
            if (Build.VERSION.SDK_INT == Build.VERSION_CODES.Q) {
                ask(Manifest.permission.ACCESS_BACKGROUND_LOCATION)
            } else {
                openForResult(Power.appDetailsIntent(context), RC_BACKGROUND_LOCATION)
                markAsked(Manifest.permission.ACCESS_BACKGROUND_LOCATION)
            }
        }

        if ("overlay" !in skip && !Overlay.isGranted(context)) {
            openForResult(overlayIntent(), RC_OVERLAY)
        }

        if ("dndAccess" !in skip && !Alerts.hasDndAccess(context)) {
            openForResult(Intent(Settings.ACTION_NOTIFICATION_POLICY_ACCESS_SETTINGS), RC_DND)
        }

        if ("fullScreenIntent" !in skip &&
            Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE &&
            !Alerts.canUseFullScreenIntent(context)
        ) {
            openForResult(fullScreenIntentSettings(), RC_FULL_SCREEN)
        }

        if ("batteryUnrestricted" !in skip && !Power.isUnrestricted(context)) {
            openForResult(Power.batterySettingsIntent(), RC_BATTERY)
        }

        if ("autostart" !in skip && Power.hasManufacturerScreen(context) && !Power.isConfirmed(context)) {
            openForResult(Power.manufacturerIntent(context), RC_AUTOSTART)
            Power.markConfirmed(context)
        }
    }

    private fun overlayIntent(): Intent =
        Intent(Settings.ACTION_MANAGE_OVERLAY_PERMISSION, Uri.parse("package:${context.packageName}"))

    private fun fullScreenIntentSettings(): Intent =
        Intent(
            "android.settings.MANAGE_APP_USE_FULL_SCREEN_INTENT",
            Uri.parse("package:${context.packageName}")
        )

    private suspend fun openSettingsFor(which: String) {
        val intent = when (which) {
            "location", "backgroundLocation", "notifications" -> Power.appDetailsIntent(context)
            "overlay" -> overlayIntent()
            "dndAccess" -> Intent(Settings.ACTION_NOTIFICATION_POLICY_ACCESS_SETTINGS)
            "batteryUnrestricted" -> Power.batterySettingsIntent()
            "fullScreenIntent" -> fullScreenIntentSettings()
            "autostart" -> Power.manufacturerIntent(context).also { Power.markConfirmed(context) }
            "notificationAccess" -> Intent(Settings.ACTION_NOTIFICATION_LISTENER_SETTINGS)
            else -> Power.appDetailsIntent(context)
        }
        openForResult(intent, RC_BATTERY)
    }

    private suspend fun ask(vararg permissions: String) = suspendCancellableCoroutine<Unit> { continuation ->
        markAsked(*permissions)
        val manager = appContext.permissions
        if (manager == null) {
            continuation.resume(Unit)
            return@suspendCancellableCoroutine
        }
        runCatching {
            manager.askForPermissions(
                PermissionsResponseListener { if (continuation.isActive) continuation.resume(Unit) },
                *permissions
            )
        }.onFailure { if (continuation.isActive) continuation.resume(Unit) }
    }

    /**
     * Settings screens cannot be awaited by any API, so the activity result —
     * which they always deliver, if only as RESULT_CANCELED — is what tells us
     * the user came back and the state is worth re-reading.
     */
    private suspend fun openForResult(intent: Intent, requestCode: Int) =
        suspendCancellableCoroutine<Unit> { continuation ->
            val activity = appContext.activityProvider?.currentActivity
            if (activity == null) {
                // No activity: still open the screen, just without a result.
                runCatching {
                    context.startActivity(intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK))
                }
                continuation.resume(Unit)
                return@suspendCancellableCoroutine
            }
            settingsRequestCode = requestCode
            settingsContinuation = { if (continuation.isActive) continuation.resume(Unit) }
            runCatching { activity.startActivityForResult(intent, requestCode) }
                .onFailure {
                    settingsContinuation = null
                    Bus.error("SETTINGS", it.message ?: "ecran de reglages introuvable")
                    if (continuation.isActive) continuation.resume(Unit)
                }
        }

    private fun applyOverrides(options: Map<String, Any?>?) {
        if (options.isNullOrEmpty()) {
            Config.invalidate()
            return
        }
        val json = JSONObject()
        options.forEach { (key, value) -> if (value != null) json.put(key, value) }
        Config.setOverrides(context, json)
    }
}
